# 1.8 Let's Code!
Mari kita lihat lebih detail kinerja dari masing-masing metode gradient descent dengan penulisan kode program. Kita akan menggunakan tool atau library Tensorflow dan Keras.
Tensorflow adalah platform Machine Learning open-source yang dikembangkan oleh Google. Platform ini menyediakan library dan sumber daya untuk memajukan penelitian di bidang Machine Learning dan memudahkan pengembang perangkat lunak dalam membangun aplikasi-aplikasi yang didukung oleh teknologi Machine Learning. Platform ini utamanya berfokus pada pengembangan dan penggunaan teknologi Deep Learning. 

Keras merupakan Application Program Interface (API) untuk Deep Learning yang ringkas, mudah dipelajari dan dapat digunakan pada platform TensorFlow. Dengan kinerja dan skalabilitas tinggi, Keras digunakan oleh organisasi dan perusahaan besar termasuk NASA, YouTube, atau Waymo. 

Pertama, import library TensorFlow dan Keras:

In [1]:
import tensorflow as tf
from tensorflow.keras import datasets
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dense 

In [2]:
!nvidia-smi

Thu Sep  4 13:54:31 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 561.19                 Driver Version: 561.19         CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                  Driver-Model | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3060 ...  WDDM  |   00000000:01:00.0 Off |                  N/A |
| N/A   48C    P8             11W /   95W |      84MiB /   6144MiB |      4%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

Selanjutnya, unduh dataset fashion_mnist. Dataset ini sudah tersedia pada TensorFlow sehingga kita bisa mengunduh dataset tersebut dengan mudah dengan perintah berikut:

In [3]:
fashion = datasets.fashion_mnist
(X_train, y_train), (X_test, y_test) = fashion.load_data()

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 5s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


Perhatikan, fungsi load_data mengambil dataset yang terdiri dari empat bagian, yaitu data dengan variabel input (data pokok) X_train dan X_test dan data dengan variabel output (data label) y_train dan y_test. Jadi, variabel input dan output dipisah ke dalam dua file.

Untuk mendapatkan performa yang baik, lakukan normalisasi nilai piksel gambar sehingga bernilai di antara 0 dan 1 (0 – 1):

In [4]:
X_train = X_train/255.0
X_test  = X_test/255.0

Siapkan model NN sesuai dengan arsitektur yang telah ditetapkan. Pada eksperimen ini, kita menggunakan NN dengan satu hidden layer yang terdiri dari 32 node dan output layer yang terdiri dari 10 node:

In [5]:
model = Sequential()
model.add(Flatten(input_shape=[28, 28]))
model.add(Dense(32, activation='relu'))
model.add(Dense(10, activation='softmax'))    

c:\Users\rafif\anaconda3\lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Terdapat dua cara untuk membuat model jaringan pada TensorFlow dan Keras: menggunakan Sequential Model dan Functional API. Eksperimen ini menggunakan Sequential Model dengan membuat model secara bertahap dengan fungsi add.

Perhatikan, ukuran input layer disesuaikan dengan dimensi gambar fashion_mnist, yaitu 28×28 dan fungsi aktivasi pada hidden layer pertama digunakan nilai default, yaitu ReLu.

Pada masalah klasifikasi dengan setiap example hanya memiliki satu kategori atau kelas (seperti pada kasus klasifikasi gambar pakaian ini) maka setiap neuron pada output layer mewakili satu kategori dengan fungsi aktivasi Softmax. Fungsi aktivasi Softmax pada output layer akan memastikan masing-masing nilai kemungkinan pada neuron-neuron tersebut berada di antara 0 dan 1 serta total nilai-nilai kemungkinan tersebut pada saat dijumlahkan adalah 1.

Lakukan kompilasi model yang telah didefinisikan di atas dengan menggunakan optimizer SGD:

In [6]:
model.compile(loss='sparse_categorical_crossentropy', optimizer='SGD', metrics=['accuracy'])

sparse_categorical_crossentropy digunakan sebagai loss function karena studi kasus ini merupakan kategorisasi fashion dengan label berbentuk angka/integer 0 – 9. Tetapi, jika label telah dilakukan pengkodean dengan one-hot encoding maka loss function categorical_crossentropy yang digunakan.

Studi kasus ini merupakan Supervised Learning dengan masalah klasifikasi. Karena terdapat 10 kategori pada data label maka ini merupakan Multiclass Classification.

Mari kita lihat struktur jaringan yang telah dikompilasi:

In [7]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 32)             │        25,120 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │           330 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 25,450 (99.41 KB)

 Trainable params: 25,450 (99.41 KB)

 Non-trainable params: 0 (0.00 B)

Perhatikan, jumlah neuron pada input layer sebesar 784 yang merupakan hasil flatten dari gambar input berukuran 28×28. Input layer tidak memiliki paramater jaringan sehingga bernilai 0. Sedangkan pada hidden dan output layer masing-masing terdapat 25.120 dan 330 paramater jaringan sehingga eksperimen ini akan melatih total 25.450 paramater. None pada Output Shape merupakan ukuran batch yang akan ditetapkan atau dihitung kemudian pada saat pelatihan.

Lakukan pelatihan model sebanyak 50 epochs dengan batch_size = 1 untuk eksperimen SGD_1, batch_size = 32 untuk eksperimen SGD_32 dan batch_size = 60000 untuk eksperimen SGD_60k. Perintah di bawah ini adalah untuk SGD_32:

In [8]:
model.fit(X_train, y_train, epochs=50, validation_split=0.2, 
batch_size=32)


Epoch 1/50
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 910us/step - accuracy: 0.6048 - loss: 1.2669 - val_accuracy: 0.7997 - val_loss: 0.6073
Epoch 2/50
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 843us/step - accuracy: 0.8068 - loss: 0.5834 - val_accuracy: 0.8208 - val_loss: 0.5281
Epoch 3/50
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 817us/step - accuracy: 0.8257 - loss: 0.5162 - val_accuracy: 0.8303 - val_loss: 0.4863
Epoch 4/50
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 827us/step - accuracy: 0.8340 - loss: 0.4774 - val_accuracy: 0.8359 - val_loss: 0.4674
Epoch 5/50
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 821us/step - accuracy: 0.8422 - loss: 0.4569 - val_accuracy: 0.8395 - val_loss: 0.4522
Epoch 6/50
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 833us/step - accuracy: 0.8469 - loss: 0.4403 - val_accuracy: 0.8468 - val_loss: 0.4366
Epoch 7/50
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 837us/step - accuracy: 0.8530 - loss: 0.4228 - val_accuracy: 0.8448 - val_loss: 0.4469
Epoch 8/50
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 792us/step - accuracy: 0.8543 -

In [9]:
model.evaluate(X_test, y_test) 

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 757us/step - accuracy: 0.8724 - loss: 0.3702


[0.371255099773407, 0.869700014591217]